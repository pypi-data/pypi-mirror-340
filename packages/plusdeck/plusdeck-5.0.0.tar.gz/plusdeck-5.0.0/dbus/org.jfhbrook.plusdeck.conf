<?xml version="1.0" encoding="UTF-8"?> <!-- -*- XML -*- -->

<!DOCTYPE busconfig PUBLIC
          "-//freedesktop//DTD D-BUS Bus Configuration 1.0//EN"
          "http://www.freedesktop.org/standards/dbus/1.0/busconfig.dtd">

<busconfig>
  <!-- Root user can own the plusdeck service -->
  <policy user="root">
    <allow own="org.jfhbrook.plusdeck"/>
    <allow send_destination="org.jfhbrook.plusdeck"/>
  </policy>

  <!-- Allow access for the "plusdeck" group -->
  <policy group="plusdeck">
    <allow send_destination="org.jfhbrook.plusdeck"/>
  </policy>
</busconfig>
