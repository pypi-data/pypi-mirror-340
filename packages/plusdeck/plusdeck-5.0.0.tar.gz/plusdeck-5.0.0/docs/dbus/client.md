# DBus Client CLI

Assuming the DBus service is running, you may interact with the service using the client CLI:

```sh
$ python3 -m plusdeck.dbus.client --help
Usage: python3 -m plusdeck.dbus.client [OPTIONS] COMMAND [ARGS]...

  Control your Plus Deck 2C Cassette Drive through dbus.

Options:
  --log-level [DEBUG|INFO|WARNING|ERROR|CRITICAL]
                                  Set the log level
  --output [text|json]            Output either human-friendly text or JSON
  --user / --default              Connect to either the user or default bus
  --help                          Show this message and exit.

Commands:
  config        Configure plusdeck.
  eject         Eject the tape
  expect        Wait for an expected state
  fast-forward  Fast-forward a tape
  pause         Pause the tape
  play          Play a tape
  rewind        Rewind a tape
  state         Get the current state
  stop          Stop the tape
  subscribe     Subscribe to state changes
```

The interface is similar to the vanilla CLI. However, there are a few differences:

1. By default, the DBus client CLI will connect to the default bus. To connect to the user session bus, set the `--user` flag. To connect to the system bus, set the `--system` flag.
2. Configuration commands do not reload the service's configuration. Instead, they will update the relevant config file, and show the differences between the file config and the service's loaded config.
3. If the config file isn't owned by the user, the client CLI will attempt to run editing commands with `sudo`.

## Installing the `plusdeck-dbus` Shim

Included in this project is `./bin/plusdeck-dbus`, a script that you can add to your PATH for convenience. If you primarily interact with the device through DBus, you may want to name this `plusdeck` on your system.
