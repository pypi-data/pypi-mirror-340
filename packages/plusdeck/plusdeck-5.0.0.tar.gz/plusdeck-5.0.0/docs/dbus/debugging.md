# Debugging Dbus

## Default Dbus Configuration

The default Dbus configuration is at `/usr/share/dbus-1/system.conf`. It may be useful to refer to this file when trying to understand what default access policies are being applied.

## Monitoring Dbus

The best tool for debugging Dbus seems to be [dbus-monitor](https://dbus.freedesktop.org/doc/dbus-monitor.1.html). To follow system bus messages, run:

```sh
sudo dbus-monitor --system
```

## Viewing Dbus Logs

You can review recent logs by checking the status of the `dbus` unit:

```sh
sudo systemctl status dbus
```

## Printing the Live Dbus Interface

I have a just task for that:

```sh
just print-iface
```
