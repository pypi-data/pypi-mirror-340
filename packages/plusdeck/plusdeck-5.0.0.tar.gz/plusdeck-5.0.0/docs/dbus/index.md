# DBus Support

The `plusdeck` library includes a DBus interface, service and client. This service allows for multitenancy on Linux - the centralized service controls the serial bus, and clients (including `python3 -m plusdeck.dbus.client`) can connect to the service.

For information on the API, visit [the API docs for dbus](../api/plusdeck.dbus/index.md):
