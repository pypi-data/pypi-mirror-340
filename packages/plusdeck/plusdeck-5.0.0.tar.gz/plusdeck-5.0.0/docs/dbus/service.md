# DBus Service

`plusdeck` includes a dbus service, which can be started either through SystemD or directly through the command line.

## Starting the Service with SystemD

`plusdeck` ships with a systemd unit that configures the service as a Dbus service. To set up the service, run:

```sh
sudo systemctl enable plusdeck
sudo systemctl start plusdeck  # optional
```

This unit will start on the `system` bus, under the root user.

## Running the Service Directly

The DBus service can be launched directly using `python3 -m plusdeck.dbus.service`:

```sh
$ python3 -m plusdeck.dbus.service --help
Usage: python3 -m plusdeck.dbus.service [OPTIONS]

  Expose the Plus Deck 2C PC Cassette Deck as a DBus service.

Options:
  --global / --no-global          Load the global config file at
                                  /etc/plusdeck.yaml (default true when
                                  called with sudo)
  -C, --config-file PATH          A path to a config file
  --log-level [DEBUG|INFO|WARNING|ERROR|CRITICAL]
                                  Set the log level
  --user / --system               Connect to either the user or system bus
  --help                          Show this message and exit.
```

In most cases, this can be called without arguments. By default, the service will listen on the `system` bus and load the global config file (`/etc/plusdeck.yml`) if launched as root; and otherwise listen on the `user` bus and load the user's config file (`~/.config/plusdeck.yml`).
