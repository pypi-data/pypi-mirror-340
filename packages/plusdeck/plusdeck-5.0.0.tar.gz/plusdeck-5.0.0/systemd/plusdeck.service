[Unit]
Description=Plus Deck 2C PC Cassette Deck
Requires=dbus.socket
After=dbus.socket

[Service]
Type=dbus
BusName=org.jfhbrook.plusdeck
ExecStart=/usr/bin/python3 -m plusdeck.dbus.service --system
Restart=on-failure

[Install]
WantedBy=multi-user.target
