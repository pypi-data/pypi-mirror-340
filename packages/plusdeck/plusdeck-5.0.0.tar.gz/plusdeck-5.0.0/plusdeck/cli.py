import asyncio
from dataclasses import asdict, dataclass, is_dataclass
from enum import Enum
import functools
import json
import logging
import os
import sys
from typing import Any, Callable, Coroutine, List, Literal, Optional, Self, TypeVar

import click
from serial.serialutil import SerialException

from plusdeck.client import Client, create_connection, State
from plusdeck.config import Config, GLOBAL_FILE

logger = logging.getLogger(__name__)

OutputMode = Literal["text"] | Literal["json"]


@dataclass
class Obj:
    """
    The main click context object. Contains options collated from parameters and the
    loaded config file.
    """

    config: Config
    global_: bool
    port: str
    output: OutputMode


LogLevel = (
    Literal["DEBUG"]
    | Literal["INFO"]
    | Literal["WARNING"]
    | Literal["ERROR"]
    | Literal["CRITICAL"]
)

STATES: List[str] = [state.name for state in State]


class PlusdeckState(click.Choice):
    """
    A Plus Deck 2C state.
    """

    name = "state"

    def __init__(self: Self) -> None:
        super().__init__(STATES)

    def convert(
        self: Self,
        value: str,
        param: Optional[click.Parameter],
        ctx: Optional[click.Context],
    ) -> State:
        choice = super().convert(value, param, ctx)

        return State[choice]


STATE = PlusdeckState()


def as_json(obj: Any) -> Any:
    """
    Convert an object into something that is JSON-serializable.
    """

    if isinstance(obj, Enum):
        return obj.name
    elif is_dataclass(obj.__class__):
        return asdict(obj)
    elif hasattr(obj, "as_dict"):
        return obj.as_dict()
    else:
        return obj


class Echo:
    """
    An abstraction for writing output to the terminal. Used to support the
    behavior of the --output flag.
    """

    mode: OutputMode = "text"

    def __call__(self: Self, obj: Any, *args, **kwargs) -> None:
        if self.mode == "json":
            try:
                click.echo(json.dumps(as_json(obj), indent=2), *args, **kwargs)
            except Exception as exc:
                logger.debug(exc)
                click.echo(json.dumps(repr(obj)), *args, **kwargs)
        else:
            if isinstance(obj, State):
                obj = obj.name

            click.echo(
                obj if isinstance(obj, str) else repr(obj),
                *args,
                **kwargs,
            )


echo = Echo()

AsyncCommand = Callable[..., Coroutine[None, None, None]]
SyncCommand = Callable[..., None]


def async_command(fn: AsyncCommand) -> SyncCommand:
    """
    Run an async command handler.
    """

    @functools.wraps(fn)
    def wrapped(*args, **kwargs) -> None:
        try:
            asyncio.run(fn(*args, **kwargs))
        except KeyboardInterrupt:
            pass

    return wrapped


def pass_client(run_forever: bool = False) -> Callable[[AsyncCommand], AsyncCommand]:
    def decorator(fn: AsyncCommand) -> AsyncCommand:
        @click.pass_obj
        @functools.wraps(fn)
        async def wrapped(obj: Obj, *args, **kwargs) -> None:
            port: str = obj.port

            try:
                client: Client = await create_connection(port)
            except SerialException as exc:
                click.echo(exc)
                sys.exit(1)

            # Giddyup!
            await fn(client, *args, **kwargs)

            # Close the client if we're done
            if not run_forever:
                client.close()

            # Await the client closing and surface any exceptions
            await client.closed

        return wrapped

    return decorator


R = TypeVar("R")


def pass_config(fn: Callable[..., R]) -> Callable[..., R]:
    @click.pass_obj
    @functools.wraps(fn)
    def wrapped(obj: Obj, *args, **kwargs) -> R:
        return fn(obj.config, *args, **kwargs)

    return wrapped


@click.group()
@click.option(
    "--global/--no-global",
    "global_",
    default=os.geteuid() == 0,
    help=f"Load the global config file at {GLOBAL_FILE} "
    "(default true when called with sudo)",
)
@click.option(
    "--config-file",
    "-C",
    envvar="PLUSDECK_CONFIG_FILE",
    type=click.Path(),
    help="A path to a config file",
)
@click.option(
    "--log-level",
    envvar="PLUSDECK_LOG_LEVEL",
    type=click.Choice(["DEBUG", "INFO", "WARNING", "ERROR", "CRITICAL"]),
    default="INFO",
    help="Set the log level",
)
@click.option(
    "--port",
    envvar="PLUSDECK_PORT",
    help="The serial port the device is connected to",
)
@click.option(
    "--output",
    type=click.Choice(["text", "json"]),
    default="text",
    help="Output either human-friendly text or JSON",
)
@click.pass_context
def main(
    ctx: click.Context,
    global_: bool,
    config_file: Optional[str],
    log_level: LogLevel,
    port: Optional[str],
    output: Optional[OutputMode],
) -> None:
    """
    Control your Plus Deck 2C tape deck.
    """

    logging.basicConfig(level=getattr(logging, log_level))

    file = None
    if config_file:
        if global_:
            logger.debug(
                "--config-file is specified, so --global flag will be ignored."
            )
        file = config_file
    elif global_:
        file = GLOBAL_FILE
    config: Config = Config.from_file(file=file)
    ctx.obj = Obj(
        config=config,
        global_=global_,
        port=port or config.port,
        output=output or "text",
    )

    echo.mode = ctx.obj.output


@main.group()
def config() -> None:
    """
    Configure plusdeck.
    """
    pass


@config.command()
@click.argument("name")
@pass_config
def get(config: Config, name: str) -> None:
    """
    Get a parameter from the configuration file.
    """

    try:
        echo(config.get(name))
    except ValueError as exc:
        echo(str(exc))
        sys.exit(1)


@config.command()
@pass_config
def show(config: Config) -> None:
    """
    Show the current configuration.
    """
    echo(config)


@config.command()
@click.argument("name")
@click.argument("value")
@pass_config
def set(config: Config, name: str, value: str) -> None:
    """
    Set a parameter in the configuration file.
    """
    try:
        config.set(name, value)
    except ValueError as exc:
        echo(str(exc))
        sys.exit(1)
    config.to_file()


@config.command()
@click.argument("name")
@pass_config
def unset(config: Config, name: str) -> None:
    """
    Unset a parameter in the configuration file.
    """
    try:
        config.unset(name)
    except ValueError as exc:
        echo(str(exc))
        sys.exit(1)
    config.to_file()


@main.group
def play() -> None:
    """
    Play a tape
    """


@play.command(name="a")
@async_command
@pass_client()
async def play_a(client: Client) -> None:
    """
    Play side A of the tape
    """

    client.play_a()


@play.command(name="b")
@async_command
@pass_client()
async def play_b(client: Client) -> None:
    """
    Play side B of the tape
    """

    client.play_b()


@main.group
def fast_forward() -> None:
    """
    Fast-forward a tape
    """


@fast_forward.command(name="a")
@async_command
@pass_client()
async def fast_forward_a(client: Client) -> None:
    """
    Fast-forward side A of the tape
    """

    client.fast_forward_a()


@fast_forward.command(name="b")
@async_command
@pass_client()
async def fast_forward_b(client: Client) -> None:
    """
    Fast-forward side B of the tape
    """

    client.fast_forward_b()


@main.group
def rewind() -> None:
    """
    Rewind a tape
    """


@rewind.command(name="a")
@async_command
@pass_client()
async def rewind_a(client: Client) -> None:
    """
    Rewind side A of the tape
    """

    client.rewind_a()


@rewind.command(name="b")
@async_command
@pass_client()
async def rewind_b(client: Client) -> None:
    """
    Rewind side B of the tape
    """

    client.rewind_b()


@main.command
@async_command
@pass_client()
async def pause(client: Client) -> None:
    """
    Pause the tape
    """

    client.pause()


@main.command
@async_command
@pass_client()
async def stop(client: Client) -> None:
    """
    Stop the tape
    """

    client.stop()


@main.command
@async_command
@pass_client()
async def eject(client: Client) -> None:
    """
    Eject the tape
    """

    client.eject()


@main.command
@click.argument("state", type=STATE)
@click.option(
    "--timeout",
    type=float,
    help="How long to wait for a state change from the Plus Deck 2C before timing out",
)
@async_command
@pass_client()
async def expect(client: Client, state: State, timeout: Optional[float]) -> None:
    """
    Wait for an expected state
    """

    async with client.session() as rcv:
        try:
            await rcv.expect(state, timeout=timeout)
        except TimeoutError:
            logger.info(f"Timed out after {timeout} seconds.")


@main.command
@click.option("--for", "for_", type=float, help="Amount of time to listen for reports")
@async_command
@pass_client(run_forever=True)
async def subscribe(client: Client, for_: Optional[float]) -> None:
    """
    Subscribe to state changes
    """

    running = True

    async def subscribe() -> None:
        async with client.session() as rcv:
            while True:
                if not running:
                    break
                try:
                    state = await rcv.get_state(timeout=1.0)
                    echo(state)
                except TimeoutError:
                    pass

    subscription = client.loop.create_task(subscribe())

    if for_ is not None:
        await asyncio.sleep(for_)
        running = False
        await subscription
        client.close()
        await client.closed
    else:
        await subscription


if __name__ == "__main__":
    main()
