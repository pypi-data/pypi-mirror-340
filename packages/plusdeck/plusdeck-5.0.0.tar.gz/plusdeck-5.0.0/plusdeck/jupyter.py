import asyncio
from dataclasses import replace

import ipywidgets as widgets
from serial.tools.list_ports import comports

from plusdeck.client import Client, Command, State
from plusdeck.config import Config

"""
ipywidgets widgets and other tools for interacting with the Plus Deck 2C PC
Cassette Deck through Jupyter.
"""


class ConfigEditor(widgets.VBox):
    """A widget for editing a Plus Deck 2C PC Cassette Deck config."""

    _config: Config
    _port_widget: widgets.SelectMultiple

    def __init__(self, config: Config):
        ports = [port.device for port in comports(include_links=True)]
        self._config = config
        self._port_widget = widgets.SelectMultiple(
            options=ports,
            value=(config.port,) if config.port in ports else ports[0:1],
            description="Serial Port",
            disabled=False,
        )
        super(ConfigEditor, self).__init__([self._port_widget])

    @property
    def value(self) -> Config:
        """The value of the current config."""

        self._config = replace(self._config, port=self._port_widget.value[0])
        return self._config


STATES = {
    State.PLAYING_A: "Play Side A",
    State.PAUSED_A: "Pause Side A",
    State.PLAYING_B: "Play Side B",
    State.PAUSED_B: "Pause Side B",
    State.FAST_FORWARDING_A: "Fast-Forward A (Rewind B)",
    State.FAST_FORWARDING_B: "Fast-Forward B (Rewind A)",
    State.STOPPED: "Stop",
    State.EJECTED: "Eject",
    State.SUBSCRIBING: "Hello",
    State.SUBSCRIBED: "Hello",
    State.UNSUBSCRIBING: "Goodbye",
    State.UNSUBSCRIBED: "Off",
}


async def player(client: Client) -> widgets.HBox:
    """Create a player widget for the Plus Deck 2C PC Cassette Deck."""

    rcv = await client.subscribe()

    status = widgets.Label(
        value="Off",
        layout={"height": "60%"},
        style={"font_family": "monospace", "font_size": "24pt"},
    )

    async def display_state():
        async for state in rcv:
            status.value = STATES[state]

    asyncio.get_running_loop().create_task(display_state())

    pause = widgets.Button(
        value=False, description="⏸️", tooltip="Pause", layout={"width": "95%"}
    )

    @pause.on_click
    def on_pause(button):
        client.send(Command.PAUSE)

    play_a = widgets.Button(
        value=False, description="▶️", tooltip="Play Side A", layout={"width": "45%"}
    )

    @play_a.on_click
    def on_play_a(button):
        client.send(Command.PLAY_A)

    play_b = widgets.Button(
        value=False, description="◀️", tooltip="Play Side B", layout={"width": "45%"}
    )

    def on_play_b(button):
        client.send(Command.PLAY_B)

    play_b.on_click(on_play_b)

    stop = widgets.Button(
        value=False, description="⏹️", tooltip="Stop", layout={"width": "95%"}
    )

    @stop.on_click
    def on_stop(button):
        client.send(Command.STOP)

    eject = widgets.Button(value=False, description="⏏️", tooltip="Eject")

    @eject.on_click
    def on_eject(button):
        client.send(Command.EJECT)

    fast_forward = widgets.Button(value=False, description="⏩", tooltip="Fast-Forward")

    @fast_forward.on_click
    def on_fast_forward(button):
        client.send(Command.FAST_FORWARD_B)

    rewind = widgets.Button(value=False, description="⏪", tooltip="Rewind")

    @rewind.on_click
    def on_rewind(button):
        client.send(Command.FAST_FORWARD_A)

    power = widgets.Button(value=False, description="⏻", tooltip="Turn On/Off")

    @power.on_click
    def on_power(button):
        client.send(Command.UNSUBSCRIBE)

    return widgets.HBox(
        [
            widgets.VBox([pause, widgets.HBox([play_b, play_a]), stop]),
            widgets.VBox([status, widgets.HBox([eject, rewind, fast_forward, power])]),
        ]
    )
