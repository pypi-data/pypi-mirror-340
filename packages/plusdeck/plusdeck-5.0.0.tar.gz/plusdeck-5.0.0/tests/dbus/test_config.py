from typing import Any, cast

import pytest

from plusdeck.config import Config, GLOBAL_FILE
from plusdeck.dbus.config import StagedConfig

cfg_cls = cast(Any, Config)


@pytest.mark.parametrize(
    "active_config,target_config",
    [
        (
            cfg_cls(file=GLOBAL_FILE, port="/dev/ttyS0"),
            cfg_cls(file=GLOBAL_FILE, port="/dev/ttyS0"),
        ),
        (
            cfg_cls(file=GLOBAL_FILE, port="/dev/ttyS0"),
            cfg_cls(file=GLOBAL_FILE, port="/dev/ttyS4"),
        ),
    ],
)
def test_staged_config_as_dict(active_config, target_config, snapshot) -> None:
    cls = cast(Any, StagedConfig)
    staged = cls(active_config=active_config, target_config=target_config)
    assert staged.as_dict() == snapshot


@pytest.mark.parametrize(
    "active_config,target_config",
    [
        (
            cfg_cls(file=GLOBAL_FILE, port="/dev/ttyS0"),
            cfg_cls(file=GLOBAL_FILE, port="/dev/ttyS0"),
        ),
        (
            cfg_cls(file=GLOBAL_FILE, port="/dev/ttyS0"),
            cfg_cls(file=GLOBAL_FILE, port="/dev/ttyS4"),
        ),
    ],
)
def test_staged_config_repr(active_config, target_config, snapshot) -> None:
    cls = cast(Any, StagedConfig)
    staged = cls(active_config=active_config, target_config=target_config)
    assert repr(staged) == snapshot
