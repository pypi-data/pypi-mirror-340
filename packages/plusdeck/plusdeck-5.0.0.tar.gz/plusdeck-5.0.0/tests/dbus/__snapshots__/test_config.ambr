# serializer version: 1
# name: test_staged_config_as_dict[active_config0-target_config0]
  dict({
    'file': dict({
      'active': '/etc/plusdeck.yaml',
      'target': '/etc/plusdeck.yaml',
      'type': None,
    }),
    'port': dict({
      'active': '/dev/ttyS0',
      'target': '/dev/ttyS0',
      'type': None,
    }),
  })
# ---
# name: test_staged_config_as_dict[active_config1-target_config1]
  dict({
    'file': dict({
      'active': '/etc/plusdeck.yaml',
      'target': '/etc/plusdeck.yaml',
      'type': None,
    }),
    'port': dict({
      'active': '/dev/ttyS0',
      'target': '/dev/ttyS4',
      'type': 'set',
    }),
  })
# ---
# name: test_staged_config_repr[active_config0-target_config0]
  '''
    file: /etc/plusdeck.yaml
    port: /dev/ttyS0
    
  '''
# ---
# name: test_staged_config_repr[active_config1-target_config1]
  '''
    file: /etc/plusdeck.yaml
  ~ port: /dev/ttyS0 ~> /dev/ttyS4
    
  '''
# ---
